[Accelerate] [Quadrature] New Quadrature Overlay

FlexMonkey opened the pull request:
A class that simplifies approximating the definite integral of a function.

This PR contains a new class, `Quadrature` that simplifies the API to `quadrature_integrate_function` and `quadrature_integrate` (see existing docs [here](https://developer.apple.com/documentation/accelerate/quadrature)).

cc: @moiseev @airspeedswift @stephentyrone

FlexMonkey committed:
[Accelerate] [Quadrature] New Quadrature Overlay

A class that simplifies approximating the definite integral of a function.

stephentyrone commented:
@swift-ci please test

stephentyrone commented on stdlib/public/Darwin/Accelerate/CMakeLists.txt:
```diff
@@ -2,19 +2,22 @@ cmake_minimum_required(VERSION 3.4.3)
 add_swift_target_library(swiftAccelerate ${SWIFT_SDK_OVERLAY_LIBRARY_BUILD_TYPES} IS_SDK_OVERLAY
-  BNNS.swift.gyb
+BNNS.swift.gyb
+Quadrature.swift
```
Can you revert the whitespace changes here? They add a little bit of noise to the simplest application of git blame if we're trying to track down a problem in the future. It looks like there should only be a single line change in this file.

stephentyrone commented on stdlib/public/Darwin/Accelerate/Quadrature.swift:
```diff
@@ -0,0 +1,258 @@
+//
+//===----------------------------------------------------------------------===//
+
+@_exported import Accelerate
```
I don't think you need `@_exported` here.

stephentyrone commented on stdlib/public/Darwin/Accelerate/Quadrature.swift:
```diff
@@ -0,0 +1,258 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
```
Copyright date needs updating.

FlexMonkey committed:
Fixes in response to PR Review.

Change `@_exported import Accelerate` to `import Accelerate`.
Correct date in copyright comment.

stephentyrone commented on stdlib/public/Darwin/Accelerate/Quadrature.swift:
```diff
@@ -0,0 +1,258 @@
+    }
+    
+    /// The quadrature instance's integration algorithm and relevant properties.
+    public var integrator: Integrator {
```
Should this actually be exposed as a `var`? I'm not sure that anyone will want to *change* the integrator or query it's value, so maybe the `init` is all we need, especially since creating a Quadrature object is very lightweight compared to actually doing the integration.

FlexMonkey replied:
Agreed. I've made that private and immutable (pending push)

FlexMonkey replied:
Fixed & Pushed.

stephentyrone commented on stdlib/public/Darwin/Accelerate/Quadrature.swift:
```diff
@@ -0,0 +1,258 @@
+    }
+    
+    /// Number of points per subinterval. Used by the QAG integrator only; other integrators ignore this value.
+    public var qagPointsPerInterval: QAGPointsPerInterval {
```
This var is probably overkill, too. I would leave it out unless someone explicitly needs it.

FlexMonkey replied:
Removed (pending push)

FlexMonkey replied:
Fixed & Pushed.

stephentyrone commented on stdlib/public/Darwin/Accelerate/Quadrature.swift:
```diff
@@ -0,0 +1,258 @@
+    }
+    
+    /// The maximum number of subintervals in the subdivision used by QAG and QAGS integrators.
+    public var maxIntervals: Int {
```
Don't think you need this unless there's a request for it.

FlexMonkey replied:
Removed (pending push)

stephentyrone commented on stdlib/public/Darwin/Accelerate/Quadrature.swift:
```diff
@@ -0,0 +1,258 @@
+    ///
+    /// - Parameter interval: The lower and upper bounds of the integration interval.
+    /// - Parameter integrand: The function to integrate. The input value is `x` that's within the interval over which the integrand is being integrated, and the output value is the corresponding value `y = integrand(x)` at those points.
+    public func integrate(over interval: ClosedRange<Double>,
```
There should also be a version of this func that exposes the vectorized callback, since you're adding all these nice vectorized vForce functions. Specifically, it would be great to expose something like:

`public func integrate(over interval: ClosedRange<Double>, integrand: (UnsafeBufferPointer<Double>, inout UnsafeMutableBufferPointer<Double>) -> ())`

so that `integrand` can be implemented in terms of the vForce and vDSP interfaces.

FlexMonkey committed:
Code Review Fixes.

* Remove mutable integrator, simply set options in init().
* Remove mutable `maxIntervals` and `qagPointsPerInterval`.
* Update tests.

karwa commented on stdlib/public/Darwin/Accelerate/Quadrature.swift:
```diff
@@ -0,0 +1,214 @@
+        case sixtyOne = 61
+    }
+    
+    public enum Result {
```
`Result` is now in the standard library. But it's probably better to make `integrate` a regular throwing function so you get error propagation and language features like `try!/try?`.

FlexMonkey replied:
I would suggest that a failure here is a domain error that doesn't warrant throwing an error. Implemented as a `Result` (pending push).

karwa commented on stdlib/public/Darwin/Accelerate/Quadrature.swift:
```diff
@@ -0,0 +1,214 @@
+    }
+    
+    public enum QAGPointsPerInterval: Int {
+        case `default` = 0
```
FYI: Raw-values do not affect layout for Swift enums. If you want something maximally-efficient (wrapping an integer) use a struct:

```
public struct QAGPointsPerInterval {
  public let points: Int
  private init(points: Int) { self.points = points }

  static let fifteen = QAGPointsPerInterval(points: 15)
  static let twentyOne = QAGPointsPerInterval(points: 21)
  // ..etc
}
```

FlexMonkey replied:
Implemented as a struct (pending push).

karwa commented on stdlib/public/Darwin/Accelerate/Quadrature.swift:
```diff
@@ -0,0 +1,214 @@
+        integrandPointer.storeBytes(of: self,
+                                    as: Quadrature.self)
+        
+        var callback = quadrature_integrate_function(
```
If you passed the address of the integrand closure in directly, the closure would no longer need to escape, calling `integrate` wouldn't mutate any state, and `Quadrature` could become a thread-safe struct which just wraps the `quadrature_integrate_options`.

FlexMonkey replied:
I'm limited here: I need to pass the integrand via `fun_arg` (which is an `UnsafeMutableRawPointer`) because a C function pointer cannot be formed from a closure that captures context. However, I've significantly tidied up that code to avoid passing an instance of the class. (Pending push)

karwa replied:
That looks nice 👍 

It's true that you can't turn a capturing closure in to a C _function_ pointer, but you can turn it in to a void-pointer that you use again from Swift (which is what you're doing). It looks like you could now drop the `@escaping` (with help from `withoutActuallyEscaping` if the compiler moans at you) and make the whole thing a struct if you wanted to.

The nice Swift interface shouldn't really cost anything. It would be unfortunate if there were unnecessary heap allocations which made it more expensive than using C directly.

FlexMonkey replied:
Great points - I've changed `Quadrature` to a struct and used `withoutActuallyEscaping`. 

Many thanks!

FlexMonkey committed:
Code Review Fixes.

* Use standard library `Result`.
* Implement `QAGPointsPerInterval` as a struct.

FlexMonkey committed:
Tidy up Passing Integrand to `quadrature_integrate_function`.

Pass the integrand closure directly to the `quadrature_integrate_function` initialiser rather than attaching as a property of the `Quadrature` class.

FlexMonkey committed:
Make `Quadrature` a struct, and remove requirement for `integrand` to be escaping.

stephentyrone commented:
Can we define some verbose and descriptive aliases of .qng, .qag, and .qags?

moiseev commented on stdlib/public/Darwin/Accelerate/Quadrature.swift:
```diff
@@ -0,0 +1,211 @@
+            }
+        }
+        
+        var errorDescription: String {
```
Shouldn't it be `public`?

FlexMonkey replied:
fixed (pending push)

moiseev commented on test/stdlib/Accelerate.swift:
```diff
@@ -10,63 +10,267 @@ import Accelerate
+            expectEqual(integralResult, legacyResult)
+            expectEqual(estimatedAbsoluteError, legacyEstimatedAbsoluteError)
+        case .failure( _):
+            expectTrue(false)
```
There is `expectationFailure()` function you can use.
```suggestion
            expectationFailure()
```

FlexMonkey replied:
fixed (pending push)

moiseev commented on test/stdlib/Accelerate.swift:
```diff
@@ -10,63 +10,267 @@ import Accelerate
+            expectEqual(integralResult, legacyResult)
+            expectEqual(estimatedAbsoluteError, legacyEstimatedAbsoluteError)
+        case .failure( _):
+            expectTrue(false)
```
```suggestion
            expectationFailure()
```

FlexMonkey replied:
fixed (pending push)

moiseev commented on test/stdlib/Accelerate.swift:
```diff
@@ -10,63 +10,267 @@ import Accelerate
+            expectEqual(integralResult, legacyResult)
+            expectEqual(estimatedAbsoluteError, legacyEstimatedAbsoluteError)
+        case .failure( _):
+            expectTrue(false)
```
```suggestion
            expectationFailure()
```

FlexMonkey replied:
fixed (pending push)

FlexMonkey commented:
> Can we define some verbose and descriptive aliases of .qng, .qag, and .qags?

Added aliases: `qng` -> `nonAdaptive`, `qag` -> `adaptive`, `qags` -> `adaptiveWithSingularities`.

(pending push)

Commit:
Code Review Changes

* Add long-form integrator algorithm aliases (update tests to use these aliases).
* Make quadrature error description public.
* New tests for error description and `QAGPointsPerInterval`.

FlexMonkey committed:
Vectorized Integrand for Quadrature

Create an alternative implementation of `integrate` that accepts an integrand of type `(_ input: UnsafeBufferPointer<Double>, _ result: UnsafeMutableBufferPointer<Double>)`.

karwa commented on stdlib/public/Darwin/Accelerate/Quadrature.swift:
```diff
@@ -0,0 +1,287 @@
+                            y: UnsafeMutablePointer<Double>
+                            ) in
+                            
+                            guard let integrand = arg?.load(as: ((UnsafeBufferPointer<Double>, UnsafeMutableBufferPointer<Double>) ->()).self) else {
```
Yikes! This is treating a closure of type `(U, V)->()` as a `(UBP<Double>, UMBP<Double>)->()`. Since the caller decides what they want U/V to be, that will be undefined behaviour if they choose some other type, like `Array<Double>`. I don't think this can be made generic.

Since this is overload is for batched processing, the element-wise version could be refactored to use it:
```swift
// <better name goes here>
func integrateVec(over interval: ClosedRange<Double>, integrand: (UBP<Double>, UMBP<Double>)->Void) -> Result<...> {
  // ...
}

func integrate(over interval: ClosedRange<Double>, integrand: (Double, inout Double)->Void) -> Result<...> {
  return integrateVec(over: interval) { src, dst in
    for i in 0..<src.count { integrand(src[i], dst[i]) } 
  }
}
```

Also, since the element-wise version is a convenience now, I personally think it would be _most_ convenient if the integrand had the signature `(Double)->Double`. There's no reason to read from the result buffer, and that's the kind of signature most mathematical functions you're likely to see in Swift will have.

FlexMonkey replied:
Indeed, yes, trying to generic-ize that function wasn't the best thing to do. I've changed the integrand signature to `integrand: (UBP<Double>, UMBP<Double>)->Void`. (pending push)

karwa replied:
As for the name: sure. I figured it might be ambiguous which one you meant in certain contexts, but it seems I misread the integrand signatures. They have different enough shapes that I think it shouldn't cause any difficulties in practice.

FlexMonkey replied:
Thanks. My change to `integrand: (UBP<Double>, UMBP<Double>)->Void` is available.

FlexMonkey committed:
Vectorized Integrand for Quadrature

Create an alternative implementation of `integrate` that accepts an integrand of type `(_ input: UnsafeBufferPointer<Double>, _ result: UnsafeMutableBufferPointer<Double>)`.

FlexMonkey committed:
Vectorized Integrand for Quadrature

Remove _ContiguousCollection - it's no longer used.

FlexMonkey committed:
Delete ContiguousCollection.swift

Not required.

moiseev commented:
@swift-ci Please test macOS platform

moiseev commented:
@swift-ci Please smoke test Linux platform

FlexMonkey committed:
Merge branch 'master' into accelerate-quadrature

FlexMonkey committed:
Refactor tests.

stephentyrone commented:
@swift-ci please test

Commit:
Merge branch 'master' into accelerate-quadrature

stephentyrone commented:
@swift-ci please smoke test and merge

stephentyrone commented:
@swift-ci please smoke test and merge

stephentyrone commented:
@swift-ci please smoke test linux
